### PR TITLE
Search books from Amazon with ISBN

### PR DESCRIPTION
**Pull request by neurogoo**
Add Item Page Magic button first searches from database based on ISBN and then searches Amazon.

### CODE REVIEW

File: library-app/src/Futurice/App/Library.hs [hunk 1]
```diff
@@ -111,17 +122,85 @@ addNewCover ctx cover = do
...
 
+makeAmazonAddress :: Ctx -> Text -> IO ByteString
+makeAmazonAddress ctx isbn = do
+    timeStamp <- formatTime defaultTimeLocale (iso8601DateFormat (Just "%H:%M:%S%EZ")) <$> getCurrentTime
+    pure $ "https://webservices.amazon.com/onca/xml?"
+        <> queryParameters timeStamp
+        <> "&Signature="
+        <> (urlEncode True . B64.encode . toBytes . hmacGetDigest . calcSign $ baseAddress timeStamp)
+  where
+    queryParameters ts = "AWSAccessKeyId=" <> TE.encodeUtf8 (cfgAmazonAccessKey cfg) <> "&"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: I'd write `T.intercalate "&" [ ... , ... , ... ]`

File: library-app/src/Futurice/App/Library.hs [hunk 2]
```diff
@@ -111,17 +122,85 @@ addNewCover ctx cover = do
...
+    calcSign = hmac (TE.encodeUtf8 $ cfgAmazonSecretKey cfg)
+    cfg = ctxConfig ctx
+
+fetchBookInformationFromAmazon :: Ctx -> Text -> IO (Maybe BookInformationByISBNResponse)
+fetchBookInformationFromAmazon ctx isbn = do
+    amazonAddress <- makeAmazonAddress ctx isbn
+    request <- N.parseRequest $ C.unpack amazonAddress
+    response <- N.httpLbs request (ctxManager ctx)
+    case X.parse $ DBL.toStrict $ N.responseBody response of
+      Left _error -> pure Nothing
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: does it make sense to log an error? i.e. work in `LogT IO`?

File: library-app/src/Futurice/App/Library.hs [hunk 3]
```diff
@@ -5,7 +5,11 @@
 module Futurice.App.Library (defaultMain) where
 
 import Codec.Picture                (DynamicImage, decodeImage)
+import Crypto.Hash                  (HMAC (..), SHA256, hmac)
```
[the remark this answers is not in the document]
Comment (neurogoo) [reply library-app/src/Futurice/App/Library.hs]: Hmm, what reason would be for changing library? The functionality would be still same?

[the remark this answers is not in the document]
Comment (phadej) [reply library-app/src/Futurice/App/Library.hs]: I'd like us to converge on `cryptohash-sha512` and `cryptohash-sha256`.

File: library-app/src/Futurice/App/Library.hs [hunk 4]
```diff
@@ -128,48 +127,53 @@ makeAmazonAddress ctx isbn = do
...
+        [ "AWSAccessKeyId=" <> TE.encodeUtf8 (cfgAmazonAccessKey cfg)
+        , "AssociateTag=" <> TE.encodeUtf8 (cfgAmazonAssociateTag cfg)
+        , "IdType=ISBN"
+        , "ItemId=" <> TE.encodeUtf8 isbn
+        , "Operation=ItemLookup"
+        , "ResponseGroup=Medium"
+        , "SearchIndex=Books"
+        , "Service=AWSECommerceService"
+        , "Timestamp=" <> TE.encodeUtf8 (T.replace ":" "%3A" $ T.pack ts)]
+    baseAddress ts = BS.intercalate "\n"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: should there be a trailing `"\n"` or not or it doesn't matter?

File: library-app/src/Futurice/App/Library.hs [hunk 5]
```diff
@@ -128,48 +127,53 @@ makeAmazonAddress ctx isbn = do
...
+              pure Nothing
+          Right ns -> pure $ do
+              items <- (findItems . X.children) ns
+              item <- (findItem . X.children) items
+              itemAttributes <- (findItemAttributes . X.children) item
+              detailPageUrl <- (findDetailPageUrl . X.children) item
+              image <- (findImage . X.children) item
+              imageUrl <- (findImageUrl . X.children) image
+              BookInformationByISBNResponse
+                  <$> (TE.decodeUtf8 <$> (getTitle . X.children) itemAttributes)
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: Sorry for not noticing right away: we have `decodeUtf8Lenient` exported from `Futurice.Prelude`

File: library-app/src/Futurice/App/Library.hs (hunk 4 above)
Comment (neurogoo) [reply library-app/src/Futurice/App/Library.hs]: There should not be trailing "\n"

File: library-app/src/Futurice/App/Library.hs [hunk 6]
```diff
@@ -101,27 +111,100 @@ getPersonioDataMap ctx = do
...
+makeAmazonAddress :: Ctx -> Text -> IO ByteString
+makeAmazonAddress ctx isbn = do
+    timeStamp <- formatTime defaultTimeLocale (iso8601DateFormat (Just "%H:%M:%S%EZ")) <$> getCurrentTime
+    pure $ "https://webservices.amazon.com/onca/xml?"
+        <> queryParameters timeStamp
+        <> "&Signature="
+        <> (urlEncode True . Base64.encode . calcSign $ baseAddress timeStamp)
+  where
+    queryParameters ts = BS.intercalate "&"
+        [ "AWSAccessKeyId=" <> TE.encodeUtf8 (cfgAmazonAccessKey cfg)
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: also `encodeUtf8`  is there, iirc. sorry again. (I'm quite sure you don't need to import TE).

File: library-app/src/Futurice/App/Library.hs [hunk 7]
```diff
@@ -101,27 +111,100 @@ getPersonioDataMap ctx = do
...
+                  <$> (decodeUtf8Lenient <$> (getTitle . X.children) itemAttributes)
+                  <*> Just isbn
+                  <*> (decodeUtf8Lenient <$> (getAuthors . X.children) itemAttributes)
+                  <*> (decodeUtf8Lenient <$> (getPublisher . X.children) itemAttributes)
+                  <*> (((take 4 . T.unpack . decodeUtf8Lenient) <$> (getPublished . X.children) itemAttributes) >>= readMaybe)
+                  <*> (decodeUtf8Lenient <$> getUrlLink detailPageUrl)
+                  <*> Just Map.empty
+                  <*> (DSAmazon . decodeUtf8Lenient <$> getUrlLink imageUrl)
+  where
+      findItems = (listToMaybe . filter (\n -> X.name n == "Items"))
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: should there be a helper function to `listToMaybe . filter (\n ->  X.name n == something)` ?

**Pull request merged**